File: src/patterns/structural/bridge/bridge.ts
// API using the dbs in different ways
// have two APIs (user and posts) and each can use different DBs

// Implementor
interface Database {
  find(query: string): string | null;
  update(query: string): string | null;
  create(query: string): string | null;
}

// ConcreteImplementor
export class MongoDB implements Database {
  find(query: string): string | null {
    return `Found ${query} in mongo`;
  }
  update(query: string): string | null {
    return `Updated ${query} in mongo`;
  }
  create(query: string): string | null {
    return `Created ${query} in mongo`;
  }
}

// ConcreteImplementor
export class PostgresDB implements Database {
  find(query: string): string | null {
    return `Found ${query} in postgres`;
  }
  update(query: string): string | null {
    return `Updated ${query} in postgres`;
  }
  create(query: string): string | null {
    return `Created ${query} in postgres`;
  }
}

// Abstraction
abstract class API {
  protected db: Database;
  constructor(db: Database) {
    this.db = db;
  }

  protected abstract update(query: string): string | null;

  protected abstract create(query: string): string | null;

  protected abstract find(query: string): string | null;
}

// RefinedAbstraction
class UserAPI extends API {
  constructor(db: Database) {
    super(db);
  }

  find(query: string): string | null {
    return this.db.find(query);
  }

  update(query: string): string | null {
    return this.db.update(query);
  }

  create(query: string): string | null {
    return this.db.create(query);
  }
}

// RefinedAbstraction
class PostsAPI extends API {
  constructor(db: Database) {
    super(db);
  }

  find(query: string): string | null {
    return this.db.find(query);
  }

  update(query: string): string | null {
    return this.db.update(query);
  }

  create(query: string): string | null {
    return this.db.create(query);
  }
}

const userApi = new UserAPI(new MongoDB());
console.log(userApi.find("user1"));

File: src/patterns/structural/bridge/bridge.go
package main

import "fmt"

// Implementation

type DB interface {
	find(string) string
	update(string) string
	delete(string) string
}

// ConcreteImplementations

type Mongo struct {

}

func (m *Mongo) find(search string) string {
	return fmt.Sprintf("Found results in mongo, searched for %v", search)
}

func (m *Mongo) update(search string) string {
	return fmt.Sprintf("Updated results in mongo, searched for %v", search)
}

func (m *Mongo) delete(search string) string {
	return fmt.Sprintf("Deleted results in mongo, searched for %v", search)
}

type Porstgres struct {

}

func (p *Porstgres) find(search string) string {
	return fmt.Sprintf("Found results in porstgres, searched for %v", search)
}

func (p *Porstgres) update(search string) string {
	return fmt.Sprintf("Updated results in porstgres, searched for %v", search)
}

func (p *Porstgres) delete(search string) string {
	return fmt.Sprintf("Deleted results in porstgres, searched for %v", search)
}

// Abstraction

type ApiAbstraction interface {
	get(string) string
	post(string) string
	delete(string) string
}

// RefinedAbstractions

type UserApi struct {
	db DB
}

func (u *UserApi) get(search string) string {
	return u.db.find(search)
}

func (u *UserApi) post(search string) string {
	return u.db.update(search)
}

func (u *UserApi) delete(search string) string {
	return u.db.delete(search)
}

// Usage

// func main(){
// 	userApi := &UserApi{db: &Mongo{}}
// 	fmt.Println(userApi.get("someUser"))
// }